### ISSUE TITLE
Deleted userdata owned by a deleted account remains linked with the template

### ISSUE THREAD

**Issue by shwstppr** [apache/cloudstack#9477]
##### ISSUE TYPE

 * Bug Report

##### COMPONENT NAME

~~~
Server
~~~

##### CLOUDSTACK VERSION

~~~
4.19 branch, most likely earlier versions
~~~

##### CONFIGURATION

NA

##### OS / ENVIRONMENT

NA

##### SUMMARY

Deleted userdata owned by a deleted account remains linked with the template. In UI it produces a 404 error. During VM deployment with the linked template it return error like,
`UserData linked to the template CentOS 5.6(64-bit) no GUI (XenServer) is not found`

##### STEPS TO REPRODUCE

~~~
1. As the admin, create a new user account
2. As the new user, register a userdata
3. As the admin, link the userdata to a template
4. As the admin, delete the new user
5. Observer userdata remains linked to the template
~~~

https://github.com/user-attachments/assets/7e97ae35-0875-4152-aaa0-1dab44e2148f

##### EXPECTED RESULTS

~~~
Maybe either an error while deleting the user account or the userdata that gets deleted is unlinked from the template
~~~

##### ACTUAL RESULTS

~~~
Delete userdata remains linked to the template
~~~

**Comment by shwstppr** [comment]
@harikrishna-patnala any thoughts on this

**Comment by harikrishna-patnala** [comment]
Yes @shwstppr seems like a valid issue.

Regarding the fix, I think we should allow linking of the userdata to the template only if both are created by that user. 
If we have to go with the fix of restricting the deletion of the user just because he/she created a userdata and linked to a template.

**Comment by DaanHoogland** [comment]
> Regarding the fix, I think we should allow linking of the userdata to the template only if both are created by that user. If we have to go with the fix of restricting the deletion of the user just because he/she created a userdata and linked to a template.

@harikrishna-patnala , that doesn't sound right. Let's discuss the scenarios, as the one described by @shwstppr is certainly not the only one that pertains to this linkage of resources:
- an account-user created a userdata entry
- another account-user created a template
- now several things may happen
1. the first user does the linking, or the second
2. the first user get deleted, or the second
3. the template gets deleted (while VMs are still running) though I am not sure of the latests guards against that.

on the one hand, removing the link might be resulting in unexpected behaviour of the VMs/template. On the other hand, a delete with cleanup should work on the account (and thus the user).

**Comment by nvazquez** [comment]
Hi @shwstppr @DaanHoogland @harikrishna-patnala - I reproduced the scenario and have observed that when the user is removed, the user data is removed as well as its been created by that user. However, the link between the template and the (removed) user data remains.



````
mysql> select t.id as template_id, t.name as template_name, u.id as userdata_id, u.name as userdata_name, u.removed as userdata_removed from vm_template t join user_data u on (t.user_data_id = u.id) where t.user_data_id = 2;
+-------------+---------------+-------------+---------------+---------------------+
| template_id | template_name | userdata_id | userdata_name | userdata_removed    |
+-------------+---------------+-------------+---------------+---------------------+
|         202 | Macchinina    |           2 | UserData-User | 2026-01-14 03:14:03 |
+-------------+---------------+-------------+---------------+---------------------+
1 row in set (0.00 sec)
````

I think we can follow a simpler approach: during the user data removal (already invoked by the remove user logic) we can search and remove the existing template links to the user data (marking  the `user_data_id` column as NULL on the `vm_template` table for the templates referencing the user data to be removed). What do you think?

**Comment by DaanHoogland** [comment]
> I think we can follow a simpler approach: during the user data removal (already invoked by the remove user logic) we can search and remove the existing template links to the user data (marking the `user_data_id` column as NULL on the `vm_template` table for the templates referencing the user data to be removed). What do you think?

Won't that leave a template vulnarable or disfunctional?

**Comment by nvazquez** [comment]
@DaanHoogland I don't think so, I think with the current state users may think the user data will still be applied but it wouldn't as its removed during the user removal.

Edit: after quick sync with @DaanHoogland we have defined to mark the user_data_id field and policy for each template as NULL and log the number of updated templates during the user data deletion within the account removal operation

**Comment by nvazquez** [comment]
Hi @DaanHoogland @shwstppr @harikrishna-patnala I have created the PR https://github.com/apache/cloudstack/pull/12439, can you please review?

**Comment by nvazquez** [comment]
@DaanHoogland @shwstppr as discussed I have closed my PR #12439 in favour of #11595
